pkg/debuginfo/store.go

```diff
@@ -263,16 +295,42 @@ func (s *Store) Upload(stream debuginfopb.DebugInfoService_UploadServer) error {
 		return status.Errorf(codes.Unknown, msg)
 	}
 
-	if err := s.metadataManager.update(ctx, buildID, metadataStateUploaded); err != nil {
-		level.Error(s.logger).Log("msg", "failed to update metadata", "err", err)
+	if err := s.metadataManager.update(ctx, buildID, objFileHash, metadataStateUploaded); err != nil {
+		err := fmt.Errorf("failed to update metadata with %s: %w", metadataStateUploaded, err)
+		return status.Error(codes.Internal, err.Error())
 	}
+
 	level.Debug(s.logger).Log("msg", "debug info uploaded", "buildid", buildID)
 	return stream.SendAndClose(&debuginfopb.UploadResponse{
 		BuildId: buildID,
 		Size:    r.size,
 	})
 }
 
+func isStale(metadataFile *metadata) bool {
+	return time.Now().Add(-15 * time.Minute).After(time.Unix(metadataFile.UploadStartedAt, 0))
+}
+
+// check returns true if the incoming object file should be uploaded.
+func check(objFile string) (bool, error) {
+	hasDWARF, err := elfutils.HasDWARF(objFile)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			// Failed to read the file, better for client to upload a new version.
+			return true, nil
+		}
+		var fe *elf.FormatError
+		if errors.As(err, &fe) {
+			if strings.Contains(fe.Error(), "bad magic number '[0 0 0 0]'") {
```

Comment:
elf expects to be a very specific magic number, no? I would not just match on `'[0 0 0 0]'`, but any `bad magic number` error.